server/src/main/java/org/opensearch/index/shard/IndexShard.java

```diff
@@ -4540,12 +4542,137 @@ public void syncSegmentsFromRemoteSegmentStore(boolean overrideLocal, boolean re
+        } finally {
+            store.decRef();
+            remoteStore.incRef();
+        }
+    }
+
+    private RemoteSegmentStoreDirectory getRemoteSegmentDirectoryForShard() {
+        FilterDirectory remoteStoreDirectory = (FilterDirectory) remoteStore.directory();
+        assert remoteStoreDirectory.getDelegate() instanceof FilterDirectory
+            : "Store.directory is not enclosing an instance of FilterDirectory";
+        FilterDirectory byteSizeCachingStoreDirectory = (FilterDirectory) remoteStoreDirectory.getDelegate();
+        final Directory remoteDirectory = byteSizeCachingStoreDirectory.getDelegate();
+        assert remoteDirectory instanceof RemoteSegmentStoreDirectory : "remoteDirectory is not an instance of RemoteSegmentStoreDirectory";
+        return ((RemoteSegmentStoreDirectory) remoteDirectory);
+    }
+
+    /**
+     * Downloads segments from given remote segment store for a specific commit.
+     * @param overrideLocal flag to override local segment files with those in remote store
+     * @param sourceRemoteSegmentDirectory RemoteSegmentDirectory Instance from which we need to sync segments
+     * @param primaryTerm Primary Term for shard at the time of commit operation for which we are syncing segments
+     * @param commitGeneration commit generation at the time of commit operation for which we are syncing segments
+     * @throws IOException if exception occurs while reading segments from remote store
+     */
+    public void syncSegmentsFromGivenRemoteSegmentStore(
+        boolean overrideLocal,
+        RemoteSegmentStoreDirectory sourceRemoteSegmentDirectory,
+        long primaryTerm,
+        long commitGeneration
+    ) throws IOException {
+        logger.info("Downloading segments from given remote segment store");
+        RemoteSegmentStoreDirectory remoteSegmentStoreDirectory = null;
+        if (remoteStore != null) {
+            remoteSegmentStoreDirectory = getRemoteSegmentDirectoryForShard();
+            remoteSegmentStoreDirectory.init();
+            remoteStore.incRef();
+        }
+        sourceRemoteSegmentDirectory.init();
+        Map<String, RemoteSegmentStoreDirectory.UploadedSegmentMetadata> uploadedSegments = sourceRemoteSegmentDirectory
+            .getSegmentsUploadedToRemoteStore(primaryTerm, commitGeneration);
+        final Directory storeDirectory = store.directory();
+        store.incRef();
+
+        try {
+            String segmentInfosSnapshotFilename = copySegmentFilesAndGetSegmentInfosFileName(storeDirectory,
+                sourceRemoteSegmentDirectory, remoteSegmentStoreDirectory, uploadedSegments, true,
+                overrideLocal);
+            if (segmentInfosSnapshotFilename != null) {
+                if (remoteSegmentStoreDirectory != null) {
+                    remoteSegmentStoreDirectory.uploadMetadata(uploadedSegments.keySet(), storeDirectory, primaryTerm, commitGeneration);
+                }
```

Comment:
can we call `afterRefresh`  functions to do this work  ? It should automatically upload segments as well as metadata .

Reply:
here, we are pulling data from remote store to local as well as remote continously. if we rely on afterRefresh call, then we will first pull all data to local and then `afterRefresh` will push all data to new remote store location, which i think could have more chances of failure in case of large shard.